include/hlslib/common/OpenCL.h

```diff
@@ -642,6 +843,87 @@ class Buffer {
     extendedPointer.param = 0;
     return extendedPointer;
   }
+
+  ExtendedMemoryPointer CreateExtendedPointer(void *hostPtr,
+                                              StorageType storageType,
+                                              uint8_t bankIndex) {
+    ExtendedMemoryPointer extendedPointer;
+    extendedPointer.obj = hostPtr;
+    extendedPointer.param = 0;
+
+    switch(storageType) {
+      case StorageType::HBM:
+        if(bankIndex >= 32)
+          ThrowRuntimeError("HBM bank index out of range. The bank index must be below 32.");
+        extendedPointer.flags = bankIndex | hbmStorageMagicNumber;
+        break;
+      case StorageType::DDR:
+        switch(bankIndex) {
+          case 0:
+            extendedPointer.flags = kMemoryBank0;
+            break;
+          case 1:
+            extendedPointer.flags = kMemoryBank1;
+            break;
+          case 2:
+            extendedPointer.flags = kMemoryBank2;
+            break;
+          case 3:
+            extendedPointer.flags = kMemoryBank3;
+            break;
+          default:
+            ThrowRuntimeError("DDR bank index out of range. The bank index must be below 4.");
+        }
+    }
+    return extendedPointer;
+  }
+#endif
+
+#ifndef HLSLIB_SIMULATE_OPENCL
+  inline void BlockOffsetsPreprocess(const std::array<size_t, 3> &blockSize, const std::array<size_t, 3> &hostBlockSizes,
+                        const std::array<size_t, 3> &deviceBlockSizes, const std::array<size_t, 3> &hostBlockOffsets, 
+                        const std::array<size_t, 3> &deviceBlockOffsets, std::array<size_t, 3> &copyBlockSizePrepared, 
+                        std::array<size_t, 2> &hostBlockSizesBytes, std::array<size_t, 2> &deviceBlockSizesBytes,
+                        std::array<size_t, 3> &hostBlockOffsetsPrepared, std::array<size_t, 3> &deviceBlockOffsetsPrepared,
+                        const size_t multiplyBy) {
+    copyBlockSizePrepared = {blockSize[0] * multiplyBy, blockSize[1], blockSize[2]};
+    hostBlockSizesBytes = {hostBlockSizes[0] * multiplyBy, hostBlockSizes[0] * hostBlockSizes[1] * multiplyBy};
+    deviceBlockSizesBytes = {deviceBlockSizes[0] * multiplyBy, deviceBlockSizes[0] * deviceBlockSizes[1] * multiplyBy};
+    hostBlockOffsetsPrepared = {hostBlockOffsets[0] * multiplyBy, hostBlockOffsets[1], hostBlockOffsets[2]};
+    deviceBlockOffsetsPrepared = {deviceBlockOffsets[0] * multiplyBy, deviceBlockOffsets[1], deviceBlockOffsets[2]};
+  }
+
+#else
+  //Is this syntax correct? (for the Iterators)
+  template <typename IteratorType1, typename IteratorType2, 
+  typename = typename std::enable_if<IsIteratorOfType<IteratorType1, T>() && IsRandomAccess<IteratorType1>()>::type,
+  typename = typename std::enable_if<IsIteratorOfType<IteratorType2, T>() && IsRandomAccess<IteratorType2>()>::type>
```

Comment:
This is not so important, since it's an internal function that we don't expect anyone to call from the outside (make sure this is clear btw), but it looks correct to me